### DISCUSSION TITLE
[4.x]: Change propagation method programmatically

### DISCUSSION THREAD

**Discussion by yannkost** [craftcms/cms#16091] [Q&A] [answered]
### Description
We would like to change the propagation method only for 1 user group (save all entries to all sites, only for this particular group). But I can't seem to write the necessary code with the EVENT_BEFORE_SAVE event.

I have following code in the event hook: 

```
Event::on(
     Entry::class,
     Element::EVENT_BEFORE_SAVE,
     function (ModelEvent $e) {
         if (ElementHelper::isDraftOrRevision($e->sender)) {
             return;
         }

         $entry = $e->sender;
         if ($entry->getSection()->handle === 'applications') {
             $currentUser = Craft::$app->getUser()->getIdentity();
             if ($currentUser) {
                 $isInGroup = $currentUser->isInGroup('pm');
                 if ($isInGroup) {
                     if ($entry instanceof Entry) {
                         $supportedSites = $entry->getSupportedSites();
                         $siteStatuses = ElementHelper::siteStatusesForElement($entry);
                         foreach (Craft::$app->getSites()->allSites as $site) {
                             $siteStatuses[$site->id] = true;
                         }
                         $entry->setEnabledForSite($siteStatuses);
                         $entry->propagateAll = true;
                         $entry->setScenario(Element::SCENARIO_LIVE);
                     }
                 }
             }
         }
     }
 );
```
Is there any way to do this programmatically? Any help greatly appreciated :)

### Steps to reproduce

1. Save an entry where the propagation method is to only save the entry to the site the element has been created on.

### Expected behavior
Having the hook code working to change the propagation method only for this particular group.

### Actual behavior

The entry gets saved without propagation.

### Craft CMS version

4.12.8

### PHP version

8.2

### Operating system and version

linux

### Database type and version

mariadb

### Installed plugins and versions

-

**Comment by brandonkelly** [comment, marked as answer]
I think the only problem is the `isDraftOrRevision()` check. When you first create a new entry, and when you are making edits to an entry, it’s usually going to be a draft, so this condition is preventing your code from working when it should.

Instead, you should be checking to make sure the entry isn’t already propagating.

Here’s updated code that should work. (I also simplified a couple things that didn’t need to be there.)

```php
use craft\elements\Entry;
use craft\events\ModelEvent;
use yii\base\Event;

Event::on(Entry::class, Entry::EVENT_BEFORE_SAVE, function(ModelEvent $event) {
    /** @var Entry $entry */
    $entry = $event->sender;

    if (
        !$entry->propagating &&
        $entry->getSection()?->handle === 'applications' &&
        Craft::$app->user->getIdentity()?->isInGroup('pm')
    ) {
        $siteStatuses = [];
        foreach (Craft::$app->sites->getAllSites() as $site) {
            $siteStatuses[$site->id] = true;
        }
        $entry->setEnabledForSite($siteStatuses);
    }
});
```

**Reply by yannkost** [reply]
Thanks a lot for the sample code. Sadly, the new create entry does not appear on the other sites. But I checked with xdebug, the code in the if statement gets executed. I suppose it just isn't possible for now.

**Reply by brandonkelly** [reply]
I tested locally and it seemed to work for me, when creating a brand new entry.

**Reply by yannkost** [reply]
hm now I'm curious: my settings for the section are: saving the entry only to the website they were created in. The default status is deactivated for all sites. When saving a new entry, it does not appear on the other sites, even when the user belongs to the group pm (as mentioned, the code gets executed in the if statement). 
Is there something I'm missing?
